src/core/hiwire.c

```diff
@@ -366,6 +366,21 @@ EM_JS_NUM(bool, hiwire_is_function, (JsRef idobj), {
   // clang-format on
 });
 
+EM_JS_NUM(bool, hiwire_is_promise, (JsRef idobj), {
+  // clang-format off
+  let obj = Module.hiwire.get_value(idobj);
+  return Object.prototype.toString.call(obj) === "[object Promise]";
```

Comment:
This looks a bit fragile. If we go this way, maybe `obj & Object.prototype..` as in https://stackoverflow.com/a/45762727/1791279

Otherwise how about just duck typing on the then method https://stackoverflow.com/a/53955664/1791279 ?

It would be also good to test this function. What would we need to do it?

Reply:
> This looks a bit fragile.  If we go this way, maybe `obj & Object.prototype...`

I am pretty much agnostic on which way to do this, if you prefer duck typing I am happy to go that way. I dropped `obj &&` because this is for type conversions code and we already test for all javascript falsey types and convert them to something else. If `obj` is `null` or `undefined`, it is a logic error somewhere else and it is better that this function should throw an error rather than silently return `false`. 

Welcome to the unhappy world of javascript: there are 10 different weird ways to do every basic task and you have to pick. (The standard library added `Array.isArray` to get rid of the "how do I tell if it's an array" mess, why not clean up these other messes by adding `Map.isMap`, `Promise.isPromise`, `Error.isError`, etc etc.)

Perhaps it's worth noting here that `Promise.resolve(x) === x` as suggested here https://stackoverflow.com/a/38339199/1942152 fails in firefox in some cases.

Reply:
> It would be also good to test this function. What would we need to do it?

That was the intended purpose of #937, though it may be better to do what I did in #1033 and define the function in pyodide.js:

https://github.com/iodide-project/pyodide/blob/d7c35d0f5972ec2fd3b591c7749d328c2cca0912/src/pyodide.js#L364

Then wrap in hiwire:
```C
EM_JS_NUM(bool, hiwire_function_supports_kwargs, (JsRef idfunc), {
  let funcstr = Module.hiwire.get_value(idfunc).toString();
  return Module.function_supports_kwargs(funcstr);
});
```
Then we can test with selenium:
```python
    for (s, res) in tests:
        s = f"function f({s}){{}}"
        selenium.run_js(
            f"return pyodide._module.function_supports_kwargs({repr(s)})"
        ) == res
```

Reply:
> This looks a bit fragile. If we go this way, maybe `obj && Object.prototype`

Actually I tested this on every falsey javascript value:
https://stackoverflow.com/questions/19839952/all-falsey-values-in-javascript
and it returns `false` for all of them so the version with `obj &&` in all cases does the same thing as the version without it.